crates/neon/src/types_impl/extract/types.rs

```diff
@@ -0,0 +1,356 @@
+
+    fn try_from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Result<Self, Self::Error>>
+    where
+        C: Context<'cx>,
+    {
+        let mut b = false;
+
+        unsafe {
+            match sys::get_value_bool(cx.env().to_raw(), v.to_local(), &mut b) {
+                sys::Status::BooleanExpected => return Ok(Err(TypeExpected::new())),
+                sys::Status::PendingException => return Err(Throw::new()),
+                status => assert_eq!(status, sys::Status::Ok),
+            }
+        }
+
+        Ok(Ok(b))
+    }
+
+    from_js!();
+}
+
+impl private::Sealed for bool {}
+
+impl<'cx> TryFromJs<'cx> for String {
+    type Error = TypeExpected<JsString>;
+
+    fn try_from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Result<Self, Self::Error>>
+    where
+        C: Context<'cx>,
+    {
+        let env = cx.env().to_raw();
+        let v = v.to_local();
+        let mut len = 0usize;
+
+        unsafe {
+            match sys::get_value_string_utf8(env, v, ptr::null_mut(), 0, &mut len) {
+                sys::Status::StringExpected => return Ok(Err(TypeExpected::new())),
+                sys::Status::PendingException => return Err(Throw::new()),
+                status => assert_eq!(status, sys::Status::Ok),
+            }
+        }
+
+        // Make room for null terminator to avoid losing a character
+        let mut buf = Vec::<u8>::with_capacity(len + 1);
+        let mut written = 0usize;
+
+        unsafe {
+            assert_eq!(
+                sys::get_value_string_utf8(
+                    env,
+                    v,
+                    buf.as_mut_ptr().cast(),
+                    buf.capacity(),
+                    &mut written,
+                ),
+                sys::Status::Ok,
+            );
+
+            debug_assert_eq!(len, written);
+            buf.set_len(len);
+
+            Ok(Ok(String::from_utf8_unchecked(buf)))
+        }
+    }
+
+    from_js!();
+}
+
+impl private::Sealed for String {}
+
+#[cfg_attr(docsrs, doc(cfg(feature = "napi-5")))]
+#[cfg(feature = "napi-5")]
+/// Extract an [`f64`] from a [`JsDate`]
+pub struct Date(pub f64);
+
+#[cfg_attr(docsrs, doc(cfg(feature = "napi-5")))]
+#[cfg(feature = "napi-5")]
+impl<'cx> TryFromJs<'cx> for Date {
+    type Error = TypeExpected<JsDate>;
+
+    fn try_from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Result<Self, Self::Error>>
+    where
+        C: Context<'cx>,
+    {
+        let mut d = 0f64;
+
+        unsafe {
+            match sys::get_date_value(cx.env().to_raw(), v.to_local(), &mut d) {
+                sys::Status::DateExpected => return Ok(Err(TypeExpected::new())),
+                sys::Status::PendingException => return Err(Throw::new()),
+                status => assert_eq!(status, sys::Status::Ok),
+            }
+        }
+
+        Ok(Ok(Date(d)))
+    }
+
+    from_js!();
+}
+
+impl private::Sealed for Date {}
+
+impl<'cx> TryFromJs<'cx> for () {
```

Comment:
What's the rationale for `()` to correspond to any JS value vs just `undefined` or `null | undefined`? I don't have a strong intuition here.

Reply:
I added a detailed comment, but the quick version is that there's a blanket impl of `FromArgs` for `T` where `T: TryFromJs`. This makes single argument functions ergonomic and removes the "`(T)` isn't a tuple, but `(T,)`" confusion.

However, it introduces an ambiguity of whether `let () = cx.args()?` is extracting unit from the first argument or is extracting zero arguments. Infallibly extracting unit removes all impact from the ambiguity.